File: day15.py
def get_hash(s):
    ret = 0
    for c in s:
        ret = (ret + ord(c)) * 17 % 256
    return ret

def part1():
    ans = 0
    with open("in.txt", "r") as file:
        input_str = file.read()
        for line in input_str.split(','):
            ans += get_hash(line)
    print(ans)

def part2():
    boxes = [[] for _ in range(256)]
    with open("in.txt", "r") as file:
        input_str = file.read()
        for line in input_str.split(','):
            if line[-1] == '-':
                line = line[:-1]
                box = get_hash(line)
                for k, (key, _) in enumerate(boxes[box]):
                    if key == line:
                        del boxes[box][k]
                        break
            else:
                key, lens = line.split("=")
                lens = int(lens)
                box = get_hash(key)
                flag = True
                for k, (existing_key, _) in enumerate(boxes[box]):
                    if existing_key == key:
                        boxes[box][k] = [key, lens]
                        flag = False
                        break
                if flag:
                    boxes[box].append([key, lens])

    ans = 0
    for i in range(256):
        for j, (_, value) in enumerate(boxes[i]):
            ans += (i + 1) * value * (j + 1)
    print(ans)

part1()
part2()

File: day15.cpp
#include <bits/stdc++.h>
using namespace std;
#define INPUT(var) for(string var; getline(cin, var), var!="";)
#define FOR(i, s, e) for (int i=s; i<e; ++i)

int gethash(string s) {
	int ret = 0;
	for (char c : s)
		ret = (ret + c) * 17 % 256;
	return ret;
}

void part1(string in) {
	int ans = 0;
	int n = in.size();
	for (int i = 0; i < n;) {
		int j = in.find(',', i);
		if (j == -1) j = n;
		string line = in.substr(i, j-i+1);
		i = j+1;
		if (i < n) line.pop_back();
		ans += gethash(line);
	}
	cout << ans << '\n';
}

void part2(string in) {
	vector<pair<string, int>> boxes[256];
	int n = in.size();
	for (int i = 0; i < n;) {
		int j = in.find(',', i);
		if (j == -1) j = n;
		string line = in.substr(i, j-i+1);
		i = j+1;
		if (i < n) line.pop_back();
		if (line.back() == '-') {
			line.pop_back();
			int box = gethash(line);
			FOR(i, 0, boxes[box].size()) {
				if (boxes[box][i].first == line) {
					boxes[box].erase(boxes[box].begin()+i);
					break;
				}
			}
		}
		else {
			int equals = line.find("=");
			string key = line.substr(0, equals);
			int lens = stoi(line.substr(equals+1));
			int box = gethash(key);
			bool flag = true;
			FOR(i, 0, boxes[box].size()) {
				if (boxes[box][i].first == key) {
					boxes[box][i].second = lens;
					flag = 0;
					break;
				}
			}
			if (flag)
				boxes[box].push_back({ key, lens });
		}
	}
	
	int ans = 0;
	FOR(i, 0, 256)
		FOR(j, 0, boxes[i].size())
			ans += (i+1) * boxes[i][j].second * (j+1);
	cout << ans << '\n';
}

int main() {
	freopen("in.txt", "r", stdin);
	string in; getline(cin, in);
	part1(in);
	part2(in);
}